test/compress/issue-1443.js

```diff
@@ -0,0 +1,53 @@
+unsafe_undefined: {
+    options = {
+        comparisons: true,
+        unsafe: true
+    }
+    mangle = {}
+    input: {
+        function f(undefined) {
+            var a;
+            return function() {
+                console.log(typeof a == "undefined");
+            }
+        }
+    }
+    expect: {
+        function f(n) {
+            var o;
+            return function() {
+                console.log(n===o);
+            }
+        }
+    }
+}
+
+keep_fnames: {
+    options = {
+        comparisons: true,
+        unsafe: true
+    }
+    mangle = {
+        keep_fnames: true
+    }
+    input: {
+        function f(undefined) {
+            return function() {
+                function n(a) {
+                    return a * a;
+                }
+                return typeof n == "undefined";
+            };
+        }
+    }
+    expect: {
+        function f(r) {
+            return function() {
+                function n(n) {
+                    return n * n;
+                }
+                return r === n;
```

Comment:
The problem I'm describing is related to the test itself and the `unsafe` typeof transform you cited above. `undefined` is an overridden variable that is no longer equivalent to `void 0`. The test in this PR accidentally produces correct output because the local function `n` will not triple equal anything passed as an argument to `f()`. 

Here's a concrete example of why the unsafe typeof transform is incorrect when undefined is overridden:
```
$ cat u.js
function f(undefined) {
    return function() {
        var n = 1;
        return typeof n == "undefined";
    };
}
console.log( f(1)() );
```
expected result:
```
$ node u.js
false
```
incorrect code generated:
```
$ uglifyjs u.js -b -c unsafe
function f(undefined) {
    return function() {
        var n = 1;
        return undefined === n;
    };
}

console.log(f(1)());
```
incorrect actual result:
```
$ uglifyjs u.js -b -c unsafe | node
true
```